File: src/renderer/worker/visualizer/smoothing.ts
import type { VisualizerSettings } from './types'

export interface Bar {
  value: number
  peak: number
  peakV: number
}

let mappedBuffer: Float32Array | null = null
let mappedBufferSize = 0

export const mapFrequency = (data: Uint8Array, numBars: number, mode: 'log' | 'linear'): Float32Array => {
  if (mappedBufferSize !== numBars) {
    mappedBuffer = new Float32Array(numBars)
    mappedBufferSize = numBars
  }
  const result = mappedBuffer!
  const len = data.length
  const usable = Math.max(16, Math.floor(len * 0.45))

  for (let i = 0; i < numBars; i++) {
    const t = i / numBars
    const binPos = mode === 'log'
      ? Math.pow(t, 0.55) * usable
      : t * usable
    const lo = Math.min(usable - 1, Math.floor(binPos))
    const hi = Math.min(usable - 1, lo + 1)
    const frac = binPos - lo
    result[i] = (data[lo] || 0) * (1 - frac) + (data[hi] || 0) * frac
  }
  return result
}

export const calcFrequencyAvg = (data: Uint8Array): number => {
  const len = data.length
  const step = len > 2048 ? Math.floor(len / 1024) : 1
  let avg = 0
  let count = 0
  for (let i = 0; i < len; i += step) {
    const v = data[i] || 0
    if (v > 5) avg += v * 1.2
    count++
  }
  avg /= count
  avg *= 1.4
  avg = avg / 255
  return avg
}

export const smoothBars = (
  bars: Bar[],
  mapped: Float32Array,
  frequencyAvg: number,
  settings: VisualizerSettings,
): void => {
  const numBars = settings.barCount
  const smoothing = settings.smoothing / 100
  const amp = settings.amplitudeScale / 100

  if (bars.length !== numBars) {
    bars.length = 0
    for (let i = 0; i < numBars; i++) {
      bars.push({ value: 0, peak: 0, peakV: 0 })
    }
  }

  const rise = 0.3 + (1 - smoothing) * 0.55
  const fall = 0.04 + (1 - smoothing) * 0.1

  for (let i = 0; i < numBars; i++) {
    const raw = mapped[i] / 255
    const boosted = raw * frequencyAvg + raw * 0.42
    const target = Math.min(1, boosted) * amp
    const b = bars[i]
    b.value += (target - b.value) * (target > b.value ? rise : fall)
    if (b.value > b.peak) {
      b.peak = b.value
      b.peakV = 0
    } else {
      b.peakV += 0.001
      b.peak = Math.max(0, b.peak - b.peakV)
    }
  }
}

File: native/visualizer/src/smoothing.rs
pub struct Smoother {
  bars: Vec<Bar>,
  mapped: Vec<f32>,
  count: u32,
}

#[derive(Clone, Copy)]
pub struct Bar {
  pub value: f32,
  pub peak: f32,
  pub peak_v: f32,
}

impl Default for Bar {
  fn default() -> Self {
    Self { value: 0.0, peak: 0.0, peak_v: 0.0 }
  }
}

impl Smoother {
  pub fn new(bar_count: u32) -> Self {
    Self {
      bars: vec![Bar::default(); bar_count as usize],
      mapped: vec![0.0; bar_count as usize],
      count: bar_count,
    }
  }

  pub fn bar_count(&self) -> u32 {
    self.count
  }

  pub fn bars(&self) -> &[Bar] {
    &self.bars
  }

  pub fn update(&mut self, freq_data: &[u8], bar_count: u32, use_log: bool, smoothing: f32, amp: f32) {
    if self.count != bar_count {
      self.count = bar_count;
      self.bars = vec![Bar::default(); bar_count as usize];
      self.mapped = vec![0.0; bar_count as usize];
    }

    let num = bar_count as usize;
    self.map_frequency(freq_data, num, use_log);
    let freq_avg = self.calc_frequency_avg(freq_data);

    let rise = 0.3 + (1.0 - smoothing) * 0.55;
    let fall = 0.04 + (1.0 - smoothing) * 0.1;

    for i in 0..num {
      let raw = self.mapped[i] / 255.0;
      let boosted = raw * freq_avg + raw * 0.42;
      let target = boosted.min(1.0) * amp;

      let b = &mut self.bars[i];
      let factor = if target > b.value { rise } else { fall };
      b.value += (target - b.value) * factor;

      if b.value > b.peak {
        b.peak = b.value;
        b.peak_v = 0.0;
      } else {
        b.peak_v += 0.001;
        b.peak = (b.peak - b.peak_v).max(0.0);
      }
    }
  }

  fn map_frequency(&mut self, data: &[u8], num_bars: usize, use_log: bool) {
    let len = data.len();
    let usable = 16usize.max(len * 45 / 100);

    for i in 0..num_bars {
      let t = i as f32 / num_bars as f32;
      let bin_pos = if use_log {
        t.powf(0.55) * usable as f32
      } else {
        t * usable as f32
      };
      let lo = (bin_pos as usize).min(usable - 1);
      let hi = (lo + 1).min(usable - 1);
      let frac = bin_pos - lo as f32;
      let v_lo = data.get(lo).copied().unwrap_or(0) as f32;
      let v_hi = data.get(hi).copied().unwrap_or(0) as f32;
      self.mapped[i] = v_lo * (1.0 - frac) + v_hi * frac;
    }
  }

  fn calc_frequency_avg(&self, data: &[u8]) -> f32 {
    let len = data.len();
    let usable = 16usize.max(len * 45 / 100);
    let step = if usable > 2048 { usable / 1024 } else { 1 };
    let mut avg = 0.0f32;
    let mut count = 0u32;
    let mut i = 0;
    while i < usable {
      let v = data[i] as f32;
      if v > 5.0 { avg += v * 1.2; }
      count += 1;
      i += step;
    }
    avg /= count as f32;
    avg *= 1.4;
    avg / 255.0
  }
}
